File: server/recentCwds.js
import fs from 'node:fs';
import fsp from 'node:fs/promises';
import os from 'node:os';
import path from 'node:path';

export const MAX_RECENT_CWDS = 15;

export function getRecentCwdsFilePath() {
  return path.join(os.homedir(), '.termi', 'recent_cwds.json');
}

export function resolveCwd(p) {
  if (!p || typeof p !== 'string') return '';
  const trimmed = p.trim();
  if (!trimmed) return '';
  const expanded = trimmed.startsWith('~')
    ? trimmed.replace(/^~/, os.homedir())
    : trimmed;
  return path.resolve(expanded);
}

export async function ensureTermiDir() {
  const dir = path.join(os.homedir(), '.termi');
  try {
    await fsp.mkdir(dir, { recursive: true });
  } catch {}
}

export async function loadRecentCwds() {
  const filePath = getRecentCwdsFilePath();
  let raw = '';
  try {
    raw = await fsp.readFile(filePath, 'utf8');
  } catch (err) {
    if (err.code === 'ENOENT') return [];
    return [];
  }

  let list;
  try {
    list = JSON.parse(raw);
    if (!Array.isArray(list)) return [];
  } catch {
    return [];
  }

  // Validate existence and prune non-existent directories
  const valid = [];
  let changed = false;
  for (const item of list) {
    if (typeof item !== 'string' || !item.trim()) {
      changed = true;
      continue;
    }
    const resolved = resolveCwd(item);
    try {
      const stat = await fsp.stat(resolved);
      if (stat.isDirectory()) {
        if (!valid.includes(resolved)) {
          valid.push(resolved);
        } else {
          changed = true;
        }
      } else {
        changed = true;
      }
    } catch {
      changed = true;
    }
  }

  if (valid.length > MAX_RECENT_CWDS) {
    valid.splice(MAX_RECENT_CWDS);
    changed = true;
  }

  if (changed) {
    await saveRecentCwds(valid);
  }

  return valid;
}

export async function saveRecentCwds(cwds) {
  await ensureTermiDir();
  const filePath = getRecentCwdsFilePath();
  const tempPath = `${filePath}.tmp.${Date.now()}`;
  const data = JSON.stringify(cwds.slice(0, MAX_RECENT_CWDS), null, 2);
  await fsp.writeFile(tempPath, data, 'utf8');
  await fsp.rename(tempPath, filePath);
}

export async function addRecentCwd(targetPath) {
  const resolved = resolveCwd(targetPath);
  if (!resolved) return loadRecentCwds();

  try {
    const stat = await fsp.stat(resolved);
    if (!stat.isDirectory()) {
      return loadRecentCwds();
    }
  } catch {
    return loadRecentCwds();
  }

  const current = await loadRecentCwds();
  const next = [resolved, ...current.filter((p) => p !== resolved)].slice(0, MAX_RECENT_CWDS);
  await saveRecentCwds(next);
  return next;
}

export async function removeRecentCwd(targetPath) {
  const resolved = resolveCwd(targetPath);
  const current = await loadRecentCwds();
  const next = current.filter((p) => p !== resolved && p !== targetPath);
  await saveRecentCwds(next);
  return next;
}

File: src-tauri/src/recent_cwds.rs
use std::path::{Path, PathBuf};
use tokio::fs;

pub const MAX_RECENT_CWDS: usize = 15;

fn get_recent_cwds_file_path() -> Option<PathBuf> {
    dirs::home_dir().map(|h| h.join(".termi").join("recent_cwds.json"))
}

pub fn resolve_cwd(target: &str) -> String {
    let trimmed = target.trim();
    if trimmed.is_empty() {
        return String::new();
    }
    let expanded = if trimmed.starts_with('~') {
        dirs::home_dir()
            .map(|h| trimmed.replacen('~', &h.to_string_lossy(), 1))
            .unwrap_or_else(|| trimmed.to_string())
    } else {
        trimmed.to_string()
    };

    let p = Path::new(&expanded);
    if let Ok(canonical) = std::fs::canonicalize(p) {
        #[cfg(windows)]
        {
            let s = canonical.to_string_lossy().to_string();
            return s.strip_prefix(r"\\?\").unwrap_or(&s).to_string();
        }
        #[cfg(not(windows))]
        {
            return canonical.to_string_lossy().to_string();
        }
    }
    expanded
}

pub async fn load_recent_cwds() -> Vec<String> {
    let path = match get_recent_cwds_file_path() {
        Some(p) => p,
        None => return Vec::new(),
    };

    let content = match fs::read_to_string(&path).await {
        Ok(c) => c,
        Err(_) => return Vec::new(),
    };

    let parsed: Vec<String> = match serde_json::from_str(&content) {
        Ok(l) => l,
        Err(_) => return Vec::new(),
    };

    let mut valid = Vec::new();
    let mut changed = false;

    for item in parsed {
        let trimmed = item.trim();
        if trimmed.is_empty() {
            changed = true;
            continue;
        }
        let resolved = resolve_cwd(trimmed);
        if let Ok(meta) = fs::metadata(&resolved).await {
            if meta.is_dir() {
                if !valid.contains(&resolved) {
                    valid.push(resolved);
                } else {
                    changed = true;
                }
            } else {
                changed = true;
            }
        } else {
            changed = true;
        }
    }

    if valid.len() > MAX_RECENT_CWDS {
        valid.truncate(MAX_RECENT_CWDS);
        changed = true;
    }

    if changed {
        let _ = save_recent_cwds(&valid).await;
    }

    valid
}

pub async fn save_recent_cwds(cwds: &[String]) -> Result<(), std::io::Error> {
    let file_path = match get_recent_cwds_file_path() {
        Some(p) => p,
        None => {
            return Err(std::io::Error::new(
                std::io::ErrorKind::NotFound,
                "Home directory not found",
            ))
        }
    };

    if let Some(parent) = file_path.parent() {
        fs::create_dir_all(parent).await?;
    }

    let slice = if cwds.len() > MAX_RECENT_CWDS {
        &cwds[..MAX_RECENT_CWDS]
    } else {
        cwds
    };

    let json = serde_json::to_string_pretty(slice)
        .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))?;

    let temp_path = file_path.with_extension(format!("tmp.{}", uuid::Uuid::new_v4()));
    fs::write(&temp_path, json).await?;
    fs::rename(&temp_path, &file_path).await?;
    Ok(())
}

pub async fn add_recent_cwd(target: &str) -> Vec<String> {
    let resolved = resolve_cwd(target);
    if resolved.is_empty() {
        return load_recent_cwds().await;
    }

    match fs::metadata(&resolved).await {
        Ok(m) if m.is_dir() => {}
        _ => return load_recent_cwds().await,
    }

    let mut list = load_recent_cwds().await;
    list.retain(|p| p != &resolved);
    list.insert(0, resolved);
    if list.len() > MAX_RECENT_CWDS {
        list.truncate(MAX_RECENT_CWDS);
    }
    let _ = save_recent_cwds(&list).await;
    list
}

pub async fn remove_recent_cwd(target: &str) -> Vec<String> {
    let resolved = resolve_cwd(target);
    let mut list = load_recent_cwds().await;
    list.retain(|p| p != target && p != &resolved);
    let _ = save_recent_cwds(&list).await;
    list
}
